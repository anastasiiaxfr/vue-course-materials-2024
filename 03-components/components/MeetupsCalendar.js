import { defineComponent } from 'vue/dist/vue.esm-bundler.js'

export default defineComponent({
  name: 'MeetupsCalendar',

  props: {
    meetups: {
      type: Array,
      required: true,
    },
  },

  template: `
    <div class="meetups-calendar">Календарь</div>
  `,
})
