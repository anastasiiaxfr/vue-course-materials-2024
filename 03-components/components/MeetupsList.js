import { defineComponent } from 'vue/dist/vue.esm-bundler.js'
import MeetupsCard from './MeetupsCard.js'

export default defineComponent({
  name: 'MeetupsList',

  components: {
    MeetupsCard,
  },

  props: {
    meetups: {
      type: Array,
      required: true,
    },
  },

  template: `
    <ul class="meetups-list">
      <li v-for="meetup in meetups" :key="meetup.id" class="meetups-list__item">
        <a :href="\`/meetups/\${meetup.id}\`" class="meetups-list__item-link" tabindex="0">
          <MeetupsCard :meetup="meetup" />
        </a>
      </li>
    </ul>
  `,
})
