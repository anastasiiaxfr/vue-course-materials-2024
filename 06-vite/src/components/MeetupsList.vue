<script setup lang="ts">
import MeetupsCard from '@/components/MeetupsCard.vue'
import type { MeetupDTO } from "@/types/meetups.types";

const props = defineProps<{
  meetups: MeetupDTO[]
}>()
</script>

<template>
  <ul class="meetups-list">
    <li v-for="meetup in meetups" :key="meetup.id" class="meetups-list__item">
      <a :href="`/meetups/${meetup.id}`" class="meetups-list__item-link" tabindex="0">
        <MeetupsCard :meetup="meetup" />
      </a>
    </li>
  </ul>
</template>

