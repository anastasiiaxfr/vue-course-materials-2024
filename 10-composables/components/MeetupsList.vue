<script setup lang="ts">
import MeetupsCard from './MeetupsCard.vue'
import type { MeetupDTO } from '../types/meetups.types.ts'

defineProps<{
  meetups: MeetupDTO[]
}>()
</script>

<template>
  <ul class="meetups-list">
    <li v-for="meetup in meetups" :key="meetup.id" class="meetups-list__item">
      <a :href="`/meetups/${meetup.id}`" class="meetups-list__item-link" tabindex="0">
        <MeetupsCard :meetup="meetup" />
      </a>
    </li>
  </ul>
</template>

<style scoped>
/* _meetups-list.css */
.meetups-list {
  margin: 0;
  padding: 0;
  list-style: none;
}

.meetups-list__item {
  margin: 0 0 32px 0;
  text-decoration: none;
}

.meetups-list__item-link {
  text-decoration: none;
  color: inherit;
}
</style>
