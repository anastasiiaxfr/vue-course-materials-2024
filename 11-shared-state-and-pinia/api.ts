import type { MeetupDTO } from './types/meetups.types.ts'

/** URL адрес API */
// export const API_URL = process.env.VITE_API_URL
export const API_URL = 'https://course-vue.javascript.ru/api'

/**
 * Получает список митапов с API
 */
export async function fetchMeetups(): Promise<MeetupDTO[]> {
  const response = await fetch(`${API_URL}/meetups`)
  const result = await response.json()
  if (!response.ok) {
    throw new Error(result.message)
  }
  return result
}

/**
 * Получает митап с API по ID
 * @param meetupId
 */
export async function fetchMeetup(meetupId: number): Promise<MeetupDTO> {
  const response = await fetch(`${API_URL}/meetups/${meetupId}`)
  const result = await response.json()
  if (!response.ok) {
    throw new Error(result.message)
  }
  return result
}

export async function login(email: string, password: string) {
  const response = await fetch(`${API_URL}/auth/login`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ email, password }),
  })
  const result = await response.json()
  if (!response.ok) {
    throw new Error(result.message)
  }
  return result
}
