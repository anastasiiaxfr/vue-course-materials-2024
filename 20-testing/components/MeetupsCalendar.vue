<script setup lang="ts">
import type { MeetupDTO } from '../types/meetups.types.ts'

defineProps<{
  meetups: MeetupDTO[]
}>()
</script>

<template>
  <div class="meetups-calendar">Календарь</div>
</template>

<style scoped></style>
