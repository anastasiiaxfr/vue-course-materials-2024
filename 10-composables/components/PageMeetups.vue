<script setup lang="ts">
import { ref, computed, onMounted } from 'vue'
import { UiContainer, UiAlert, UiRadioGroup, UiButton } from '@shgk/vue-course-ui'
import UiButtonGroup from './UiButtonGroup.vue'
import MeetupsList from './MeetupsList.vue'
import MeetupsCalendar from './MeetupsCalendar.vue'
import type { Ref } from 'vue'
import type { MeetupDTO } from '../types/meetups.types.ts'
import { fetchMeetups } from '../api.ts'

const view: Ref<'list' | 'calendar'> = ref('list')
const filters = ref({
  date: 'all',
  participation: 'all',
  query: '',
})

const meetups: Ref<MeetupDTO[] | null> = ref(null)

const filteredMeetups = computed(() => {
  if (!meetups.value) {
    return []
  }

  const dateFilter = (meetup: MeetupDTO) =>
    filters.value.date === 'all' ||
    (filters.value.date === 'past' && new Date(meetup.date) <= new Date()) ||
    (filters.value.date === 'future' && new Date(meetup.date) > new Date())

  const participationFilter = (meetup: MeetupDTO) =>
    filters.value.participation === 'all' ||
    (filters.value.participation === 'organizing' && meetup.organizing) ||
    (filters.value.participation === 'attending' && meetup.attending)

  const searchFilter = (meetup: MeetupDTO) =>
    [meetup.title, meetup.description, meetup.place, meetup.organizer]
      .join(' ')
      .toLowerCase()
      .includes(filters.value.query.toLowerCase())

  return meetups.value.filter(
    (meetup: MeetupDTO) => dateFilter(meetup) && participationFilter(meetup) && searchFilter(meetup),
  )
})

const dateFilterOptions = [
  { value: 'all', label: 'Все' },
  { value: 'past', label: 'Прошедшие' },
  { value: 'future', label: 'Ожидаемые' },
] as const

onMounted(async () => {
  try {
    meetups.value = await fetchMeetups()
  } catch (error) {
    console.error(error)
  }
})

// Называем переменную ref также, как template ref (атрибут ref в шаблоне)
// В его значении будет находиться ссылка на DOM элемент
const inputElement = ref<HTMLInputElement | null>(null)
// Vue 3.5+ - новый API - useTemplateRef
// Переменную можно назвать, как угодно
// const inputElement = useTemplateRef<HTMLInputElement>('inputElement')

function focusInput() {
  inputElement.value?.focus()
}
</script>

<template>
  <UiContainer>
    <div class="filters-panel">
      <div class="filters-panel__col">
        <UiRadioGroup v-model="filters.date" :options="dateFilterOptions" name="date" />
      </div>

      <div class="filters-panel__col">
        <div class="form-group form-group--inline">
          <div class="input-group input-group_icon input-group_icon-left">
            <div class="input-group__icon">
              <svg
                fill="none"
                height="24"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
                class="icon"
              >
                <g stroke="#8ca0cf" stroke-linecap="round" stroke-linejoin="round" stroke-width="2">
                  <path
                    d="m11 19c4.4183 0 8-3.5817 8-8 0-4.41828-3.5817-8-8-8-4.41828 0-8 3.58172-8 8 0 4.4183 3.58172 8 8 8z"
                  />
                  <path d="m20.9999 21-4.35-4.35" />
                </g>
              </svg>
            </div>
            <input
              ref="inputElement"
              class="form-control form-control--rounded form-control--sm"
              aria-label="Поиск"
              placeholder="Поиск"
              type="search"
              v-model.trim="filters.query"
            />
          </div>
        </div>
        <div class="form-group form-group--inline">
          <UiButtonGroup v-model:view="view" />
        </div>
      </div>

      <div class="filters-panel__col">
        <div class="form-group form-group--inline">
          <UiButton @click="focusInput">Focus</UiButton>
        </div>
      </div>
    </div>

    <template v-if="meetups !== null">
      <template v-if="filteredMeetups!.length">
        <MeetupsList v-if="view === 'list'" :meetups="filteredMeetups" />
        <MeetupsCalendar v-else-if="view === 'calendar'" :meetups="filteredMeetups" />
      </template>
      <UiAlert v-else>Митапов по заданным условиям не найдено...</UiAlert>
    </template>
    <UiAlert v-else text="Загрузка..." />
  </UiContainer>
</template>

<style scoped>
.filters-panel {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin: 40px 0 32px 0;
}

.filters-panel__col {
  display: flex;
  flex-direction: row;
  align-items: flex-start;
}

@media all and (max-width: 767px) {
  .filters-panel {
    flex-direction: column;
  }

  .filters-panel__col + .filters-panel__col {
    margin-top: 16px;
  }
}
</style>
