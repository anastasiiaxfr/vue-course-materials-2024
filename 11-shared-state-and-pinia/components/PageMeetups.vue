<script setup lang="ts">
import { ref, computed, onMounted } from 'vue'
import { UiContainer, UiAlert, UiRadioGroup, UiInput, UiFormGroup, UiIcon } from '@shgk/vue-course-ui'
import UiButtonGroup from './UiButtonGroup.vue'
import MeetupsList from './MeetupsList.vue'
import MeetupsCalendar from './MeetupsCalendar.vue'
import type { Ref } from 'vue'
import type { MeetupDTO } from '../types/meetups.types.ts'
import { fetchMeetups } from '../api.ts'

const view: Ref<'list' | 'calendar'> = ref('list')
const filters = ref({
  date: 'all',
  participation: 'all',
  query: '',
})

const meetups: Ref<MeetupDTO[] | null> = ref(null)

const filteredMeetups = computed(() => {
  if (!meetups.value) {
    return []
  }

  const dateFilter = (meetup: MeetupDTO) =>
    filters.value.date === 'all' ||
    (filters.value.date === 'past' && new Date(meetup.date) <= new Date()) ||
    (filters.value.date === 'future' && new Date(meetup.date) > new Date())

  const participationFilter = (meetup: MeetupDTO) =>
    filters.value.participation === 'all' ||
    (filters.value.participation === 'organizing' && meetup.organizing) ||
    (filters.value.participation === 'attending' && meetup.attending)

  const searchFilter = (meetup: MeetupDTO) =>
    [meetup.title, meetup.description, meetup.place, meetup.organizer]
      .join(' ')
      .toLowerCase()
      .includes(filters.value.query.toLowerCase())

  return meetups.value.filter(
    (meetup: MeetupDTO) => dateFilter(meetup) && participationFilter(meetup) && searchFilter(meetup),
  )
})

const dateFilterOptions = [
  { value: 'all', label: 'Все' },
  { value: 'past', label: 'Прошедшие' },
  { value: 'future', label: 'Ожидаемые' },
] as const

onMounted(async () => {
  try {
    meetups.value = await fetchMeetups()
  } catch (error) {
    console.error(error)
  }
})
</script>

<template>
  <UiContainer>
    <div class="filters-panel">
      <div class="filters-panel__col">
        <UiRadioGroup v-model="filters.date" :options="dateFilterOptions" name="date" />
      </div>

      <div class="filters-panel__col">
        <UiFormGroup inline>
          <UiInput v-model.trim="filters.query" aria-label="Поиск" placeholder="Поиск!" type="search" rounded small>
            <template #left-icon>
              <UiIcon icon="search" />
            </template>
          </UiInput>
        </UiFormGroup>

        <UiFormGroup inline>
          <UiButtonGroup v-model:view="view" />
        </UiFormGroup>
      </div>
    </div>

    <template v-if="meetups !== null">
      <template v-if="filteredMeetups!.length">
        <MeetupsList v-if="view === 'list'" :meetups="filteredMeetups" />
        <MeetupsCalendar v-else-if="view === 'calendar'" :meetups="filteredMeetups" />
      </template>
      <UiAlert v-else>Митапов по заданным условиям не найдено...</UiAlert>
    </template>
    <UiAlert v-else text="Загрузка..." />
  </UiContainer>
</template>

<style scoped>
.filters-panel {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin: 40px 0 32px 0;
}

.filters-panel__col {
  display: flex;
  flex-direction: row;
  align-items: flex-start;
}

@media all and (max-width: 767px) {
  .filters-panel {
    flex-direction: column;
  }

  .filters-panel__col + .filters-panel__col {
    margin-top: 16px;
  }
}
</style>
