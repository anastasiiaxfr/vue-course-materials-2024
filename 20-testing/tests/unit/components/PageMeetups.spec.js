import { vi, describe, it, expect } from 'vitest'
import { config, flushPromises, mount, shallowMount } from '@vue/test-utils'
import { nextTick } from 'vue'
import PageMeetups from '../../../components/PageMeetups.vue'
import meetupsFixture from '../../__fixtures__/meetups.json'
import { fetchMeetups } from '../../../api.ts'
import UiButtonGroup from '../../../components/UiButtonGroup.vue'
import MeetupsCalendar from '../../../components/MeetupsCalendar.vue'
import MeetupsList from '../../../components/MeetupsList.vue'

// Подменяем fetchMeetups на заглушку
// Альтернатива: msw.io
vi.mock('../../../api.ts')
fetchMeetups.mockResolvedValue(meetupsFixture)

// По умолчанию у заглушек не рендерится содержимое (слот по умолчанию)
// Это бывает не удобно - включим глобально
config.global.renderStubDefaultSlot = true

describe('PageMeetups', () => {
  it('should match snapshot', async () => {
    const wrapper = shallowMount(PageMeetups)
    await flushPromises()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should request meetups by fetchMeetups on mount', async () => {
    shallowMount(PageMeetups)
    await flushPromises()
    expect(fetchMeetups).toHaveBeenCalled()
  })

  // Unit-тест - тестируем взаимодействие с другими компонентами
  it('should render meetups by MeetupsCalendar on switch to calendar view', async () => {
    const wrapper = shallowMount(PageMeetups)
    await flushPromises()
    const uiButtonGroup = wrapper.findComponent(UiButtonGroup)
    uiButtonGroup.vm.$emit('update:view', 'calendar')
    await nextTick()
    expect(wrapper.findComponent(MeetupsList).exists()).toBe(false)
    expect(wrapper.findComponent(MeetupsCalendar).exists()).toBe(true)
    expect(wrapper.findComponent(MeetupsCalendar).props('meetups')).toEqual(meetupsFixture)
  })

  // Интеграционный тест - тестируем полный компонент
  it('should render meetups calendar on switch to calendar view', async () => {
    const wrapper = mount(PageMeetups)
    await flushPromises()
    const calendarButton = wrapper.get('[aria-label="Календарь"]')
    await calendarButton.trigger('click')
    expect(wrapper.find('.meetups-list').exists()).toBe(false)
    expect(wrapper.find('.meetups-calendar').exists()).toBe(true)
  })
})
