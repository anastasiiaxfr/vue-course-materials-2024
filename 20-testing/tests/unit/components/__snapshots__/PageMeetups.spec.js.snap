// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageMeetups > should match snapshot 1`] = `
"<ui-container-stub data-v-28f33376="">
  <div data-v-28f33376="" class="filters-panel">
    <div data-v-28f33376="" class="filters-panel__col">
      <ui-radio-group-stub data-v-28f33376="" modelvalue="all" options="[object Object],[object Object],[object Object]" name="date"></ui-radio-group-stub>
    </div>
    <div data-v-28f33376="" class="filters-panel__col">
      <div data-v-28f33376="" class="form-group form-group--inline">
        <div data-v-28f33376="" class="input-group input-group_icon input-group_icon-left">
          <div data-v-28f33376="" class="input-group__icon"><svg data-v-28f33376="" fill="none" height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg" class="icon">
              <g data-v-28f33376="" stroke="#8ca0cf" stroke-linecap="round" stroke-linejoin="round" stroke-width="2">
                <path data-v-28f33376="" d="m11 19c4.4183 0 8-3.5817 8-8 0-4.41828-3.5817-8-8-8-4.41828 0-8 3.58172-8 8 0 4.4183 3.58172 8 8 8z"></path>
                <path data-v-28f33376="" d="m20.9999 21-4.35-4.35"></path>
              </g>
            </svg></div><input data-v-28f33376="" class="form-control form-control--rounded form-control--sm" aria-label="Поиск" placeholder="Поиск" type="search">
        </div>
      </div>
      <div data-v-28f33376="" class="form-group form-group--inline">
        <ui-button-group-stub data-v-28f33376="" view="list"></ui-button-group-stub>
      </div>
    </div>
  </div>
  <meetups-list-stub data-v-28f33376="" meetups="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"></meetups-list-stub>
</ui-container-stub>"
`;
